File: rust/src/day10.rs
use std::collections::HashMap;

static INPUT: &str = include_str!("../../inputs/day10.txt");

pub fn run() {
    let numbers = transform_input(INPUT);

    println!("Part 1: {}", part1(&numbers));
    println!("Part 2: {}", part2(&numbers));
}

fn transform_input(input: &'static str) -> Vec<u64> {
    let mut numbers: Vec<u64> = input.lines().map(|x| x.parse().unwrap()).collect();

    numbers.sort_unstable();
    numbers.insert(0, 0);
    numbers.push(numbers.last().unwrap() + 3);

    numbers
}

fn part1(numbers: &[u64]) -> usize {
    let diffs: Vec<u64> = numbers.windows(2).map(|x| x[1] - x[0]).collect();

    let off_ones = diffs.iter().filter(|x| **x == 1).count();
    let off_threes = diffs.iter().filter(|x| **x == 3).count();

    off_ones * off_threes
}

fn part2(numbers: &[u64]) -> u64 {
    fn calc_arrangements(numbers: &[u64], idx: usize, memo_table: &mut HashMap<usize, u64>) -> u64 {
        // Base case.
        if idx == numbers.len() - 2 {
            return 1;
        }

        if let Some(num_paths) = memo_table.get(&idx) {
            return *num_paths;
        };

        // Recursive case.

        let mut sum = 0;
        let cur = numbers[idx];

        if let Some(true) = numbers.get(idx + 3).map(|&x| x <= cur + 3) {
            sum += calc_arrangements(numbers, idx + 3, memo_table);
        }
        if let Some(true) = numbers.get(idx + 2).map(|&x| x <= cur + 2) {
            sum += calc_arrangements(numbers, idx + 2, memo_table);
        }

        sum += calc_arrangements(numbers, idx + 1, memo_table);
        memo_table.insert(idx, sum);
        sum
    };

    calc_arrangements(numbers, 0, &mut HashMap::new())
}

File: python/src/day10.py
#!/usr/bin/env python

from collections import defaultdict
from pathlib import Path
from typing import Generator, List

INPUT_FILE = Path.cwd().parent / "inputs" / "day10.txt"


def transform_input(input: str) -> List[int]:
    numbers = sorted([int(x) for x in input.splitlines()])
    return [0, *numbers, numbers[-1] + 3]


def part1(numbers: List[int]) -> int:
    diffs = [numbers[i] - numbers[i - 1] for i in range(1, len(numbers))]
    return diffs.count(1) * diffs.count(3)


def part2(numbers: List[int]) -> int:
    num_paths = defaultdict(int)
    num_paths[0] = 1

    for window in windows(numbers, 4):
        for prev in window[:-1]:
            if window[-1] - prev <= 3:
                num_paths[window[-1]] += num_paths[prev]

    return num_paths[numbers[-1]]


def windows(list_: List, size: int) -> Generator[List, None, None]:
    for i in range(len(list_) + 1):
        yield list_[max(0, i - size) : i]


if __name__ == "__main__":
    with INPUT_FILE.open("r") as f:
        numbers = transform_input(f.read())

    print(f"Part 1: {part1(numbers)}")
    print(f"Part 2: {part2(numbers)}")
